File: easy/valid_parentheses.rb
# @param {String} s
# @return {Boolean}
def is_valid(s)
  arr = []
  s.split('').each do |c|
    if arr.length == 0
      arr.push c
      next
    end
    case c
    when ')'
      if arr.last == '('
        arr.pop
      else
        arr.push c
      end
    when ']'
      if arr.last == '['
        arr.pop
      else
        arr.push c
      end
    when '}'
      if arr.last == '{'
        arr.pop
      else
        arr.push c
      end
    else
      arr.push c
    end
  end
  arr.length == 0
end

File: easy/valid_parentheses.py
class Solution(object):
  def isValid(self, s):
    """
    :type s: str
    :rtype: bool
    """
    arr = []
    for i in range(len(s)):
      if len(arr) == 0:
        arr.append(s[i])
        continue
      if s[i] == ')':
        if arr[-1] == '(':
          arr.pop()
        else:
          arr.append(s[i])
      elif s[i] == ']':
        if arr[-1] == '[':
          arr.pop()
        else:
          arr.append(s[i])
      elif s[i] == '}':
        if arr[-1] == '{':
          arr.pop()
        else:
          arr.append(s[i])
      else:
        arr.append(s[i])
    if len(arr) == 0:
      return True
    else:
      return False
